orchagent/watermarkorch.cpp

```diff
@@ -0,0 +1,224 @@
+#include "watermarkorch.h"
+#include "sai_serialize.h"
+#include "portsorch.h"
+#include "notifier.h"
+#include "converter.h"
+
+#define DEFAULT_TELEMETRY_INTERVAL 120
+
+#define CLEAR_PG_HEADROOM_REQUEST "PG_HEADROOM"
+#define CLEAR_PG_SHARED_REQUEST "PG_SHARED"
+#define CLEAR_QUEUE_SHARED_UNI_REQUEST "Q_SHARED_UNI"
+#define CLEAR_QUEUE_SHARED_MULTI_REQUEST "Q_SHARED_MULTI"
+
+extern PortsOrch *gPortsOrch;
+
+
+WatermarkOrch::WatermarkOrch(DBConnector *db, const string tableName):
+    Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+
+    m_countersDb = make_shared<DBConnector>(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
+    m_appDb = make_shared<DBConnector>(APPL_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
+    m_countersTable = make_shared<Table>(m_countersDb.get(), COUNTERS_TABLE);
+    m_periodicWatermarkTable = make_shared<Table>(m_countersDb.get(), PERIODIC_WATERMARKS_TABLE);
+    m_persistentWatermarkTable = make_shared<Table>(m_countersDb.get(), PERSISTENT_WATERMARKS_TABLE);
+    m_userWatermarkTable = make_shared<Table>(m_countersDb.get(), USER_WATERMARKS_TABLE);
+
+    m_clearNotificationConsumer = new swss::NotificationConsumer(
+            m_appDb.get(),
+            "WATERMARK_CLEAR_REQUEST");
+    auto clearNotifier = new Notifier(m_clearNotificationConsumer, this, "WM_CLEAR_NOTIFIER");
+    Orch::addExecutor(clearNotifier);
```

Comment:
For user issued clear, why not just directly clear through the CLI but signal through an APPL_DB event channel, and delegate to WatermarkOrch to do so? What is the benefit here?

Reply:
This is the design we agreed on.
https://github.com/Azure/SONiC/blob/gh-pages/doc/watermarks_HLD.md

Reply:
This was needed in the draft design of the feature, in current implementation it doesn't give any advantages.